pyquil/api/_quantum_computer.py

```diff
@@ -411,17 +417,34 @@ def run_and_measure(self, program: Program, trials: int) -> Dict[int, np.ndarray
         :return: A dictionary keyed by qubit index where the corresponding value is a 1D array of
             measured bits.
         """
+
+        def unwrap_qubit(q: QubitDesignator) -> int:
+            if isinstance(q, Qubit):
+                return q.index
+            elif isinstance(q, int):
+                return q
+            else:
+                raise TypeError(
+                    "Cannot run and measure program with " f"unaddressed QubitPlaceholders: {q}"
+                )
+
         program = program.copy()
         validate_supported_quil(program)
         ro = program.declare("ro", "BIT", len(self.qubits()))
-        for i, q in enumerate(self.qubits()):
+        measure_all = isinstance(self.qam, QVM) and self.qam.noise_model is not None
+        qubits_to_measure = (
+            self.qubits() if not measure_all else map(unwrap_qubit, program.get_qubits())
+        )
+        for i, q in enumerate(qubits_to_measure):
             program.inst(MEASURE(q, ro[i]))
         program.wrap_in_numshots_loop(trials)
         executable = self.compile(program)
         bitstring_array = self.run(executable=executable)
         bitstring_dict = {}
-        for i, q in enumerate(self.qubits()):
+        for i, q in enumerate(qubits_to_measure):
```

Comment:
`qubits_to_measure` may be an iterator (`map`) and thus exhausted at this point - it will always be empty. Likewise below in `set(qubits_to_measure)`. If tests are passing, that indicates we need a little more coverage.

Reply:
He calls `set()` on the return of the `map()` so it shouldn't be a generator at that point.